File: snippets/fsharp/System/NullReferenceException/Overview/nullreturn2a.fs
module nullreturn2a

// <Snippet5>
open System

[<AllowNullLiteral>]
type Person(firstName) =
    member _.FirstName = firstName

    static member AddRange(firstNames) =
        Array.map Person firstNames

let persons = 
    [| "Abigail"; "Abra"; "Abraham"; "Adrian"
       "Ariella"; "Arnold"; "Aston"; "Astor" |]
    |> Person.AddRange

let nameToFind = "Robert"
let found = Array.Find(persons, fun p -> p.FirstName = nameToFind)

if found <> null then
    printfn $"{found.FirstName}"
else 
    printfn $"{nameToFind} not found."

// Using F#'s Array.tryFind function
// This does not require a null check or [<AllowNullLiteral>]
let found2 = 
    persons |> Array.tryFind (fun p -> p.FirstName = nameToFind)

match found2 with
| Some firstName ->
    printfn $"{firstName}"
| None ->
    printfn $"{nameToFind} not found."

// The example displays the following output:
//        Robert not found.
//        Robert not found.
// </Snippet5>

File: snippets/csharp/System/NullReferenceException/Overview/nullreturn2a.cs
using System;

namespace NullReturn2
{
    // <Snippet5>
    public class NullReturn2Example
    {
        public static void Main()
        {
            Person[] persons = Person.AddRange([ "Abigail", "Abra",
                                          "Abraham", "Adrian", "Ariella",
                                          "Arnold", "Aston", "Astor" ]);
            string nameToFind = "Robert";
            Person found = Array.Find(persons, p => p.FirstName == nameToFind);
            if (found != null)
                Console.WriteLine(found.FirstName);
            else
                Console.WriteLine("{0} not found.", nameToFind);
        }
    }

    public class Person
    {
        public static Person[] AddRange(string[] firstNames)
        {
            Person[] p = new Person[firstNames.Length];
            for (int ctr = 0; ctr < firstNames.Length; ctr++)
                p[ctr] = new Person(firstNames[ctr]);

            return p;
        }

        public Person(string firstName)
        {
            FirstName = firstName;
        }

        public string FirstName;
    }
    // The example displays the following output:
    //        Robert not found
    // </Snippet5>
}
